### ISSUE TITLE
Define elas import procedure

### ISSUE THREAD

**Issue by renardeau** [dimitrydhondt/inner-gerbil#19]
Import data from elas postgres database
Starting with import of users to parties

**Comment by renardeau** [comment]
Starting with write of a test for PUT of a party

**Comment by renardeau** [comment]
Mocha test on PUT parties ok
Import approach: load elas data from csv file (generated by elas export) and push a put on API for each line.
Issue: we will not be able to port user passwords this way, unless we apply same verification method: no passwords are stored on elas, only the result of a hash on it

We'll also have to add extra contact information: current contactdetails table does not contain phone or email

**Comment by dimitrydhondt** [comment]
We could also adapt the authentication method to support the hashes from elas.

**Comment by dimitrydhondt** [comment]
Are we going to migrate messages as well ?

**Comment by eeemarv** [comment]
* The authentication method of eLAS is weak and not good: MD5, SHA-1 or SHA-512 (if it matches one of those) without salt. Then better put in place a decent authentication right away. Users can reset themselves their password. OR you can support the eLAS authentication for a limited time (one year) for migrated groups: if a user signs in, then the password is stored in the new way. 
* But, anyway, probably many users will still have to reset their password because the "login" field is going to be dropped, isn't it. (Users store the login in the browser) I plan to drop the 'login' field some time soon in eLAS-Heroku. https://github.com/eeemarv/elas-heroku/issues/18
* Exporting CSV from eLAS -> is it reliable?

**Comment by renardeau** [comment]
Added import of messages.
A creation of a message requires the retrieval of the appropriate author. The elas export only contains a letscode, which is not available in inner gerbil.
So, I updated the user import to store the letscode as the 'alias' field and lookup the author based on this alias. Not sure if this is a proper usage of alias, thought

**Comment by eeemarv** [comment]
I would say, better import from the database. For sure there is much more information missing. e.i. the maxlimit of users is also missing.

**Comment by eeemarv** [comment]
Column    |            Type             |                       Modifiers                        
--------------+-----------------------------+--------------------------------------------------------
 id           | integer                     | not null default nextval('las.users_id_seq'::regclass)
 cdate        | timestamp without time zone | 
 mdate        | timestamp without time zone | 
 creator      | integer                     | not null default 0
 comments     | character varying(100)      | 
 hobbies      | text                        | not null
 name         | character varying(50)       | not null default ''::character varying
 birthday     | date                        | 
 letscode     | character varying(20)       | not null default ''::character varying
 postcode     | character varying(6)        | not null default ''::character varying
 login        | character varying(50)       | not null default ''::character varying
 cron_saldo   | boolean                     | 
 password     | character varying(150)      | 
 accountrole  | character varying(20)       | not null default ''::character varying
 status       | integer                     | not null default 0
 saldo        | integer                     | not null default 0
 lastlogin    | timestamp without time zone | not null default now()
 minlimit     | integer                     | not null default 0
 maxlimit     | integer                     | 
 fullname     | character varying(100)      | 
 PictureFile  | character varying(128)      | 
 presharedkey | character varying(80)       | 
 pwchange     | boolean                     | 
 locked       | boolean                     | 
 lang         | character varying(5)        | not null
 adate        | timestamp without time zone | 
 ostatus_id   | character varying(50)       | 
 pubkey       | text                        | 
 privkey      | text                        | 
 admincomment | character varying(200)      | 
Indexes:
    "users_id_pkey" PRIMARY KEY, btree (id)
    "users_cron_saldo" btree (cron_saldo)

**Comment by renardeau** [comment]
I should probably rephrase :-)
"A creation of a message requires the retrieval of the appropriate author. **The only usable field in elas (and elas export) is the letscode**, which is not available in inner gerbil."

All those other fields like min/max limit, balance are also still missing in inner gerbil

It felt convenient to use the CSV format and the export function looked like a good starting point. But nothing prevents us from enriching that CSV with extra data or switching to another format. However, I'd rather stick to import through the API as it prevents data inconsistencies

**Comment by eeemarv** [comment]
I opened an issue for support for letscodes #24

**Comment by renardeau** [comment]
Updated import to also create partyrelation when importing a user. The new code field is used to store the letscode.
Note: There is still an issue on PUT of a messagerelation: the balance field is ignored by insert operation

**Comment by dimitrydhondt** [comment]
Hej Bernard, I was thinking the folowing : when you import the transactions, you should have a correct balance. So there should be no need to set the balance on a partyrelation. Perhaps I'm missing something, but it seems a logical approach to me ? An eLAS instance should have started out with all account 0, and then have registered all transaction, no ?

Perhaps it is not this simple, but then I would suggest registering an initial transaction on the group to set the start balance of accounts ?

**Comment by eeemarv** [comment]
Indeed, the balance data is not primary data, but only a result from the transactions.

**Comment by renardeau** [comment]
Makes sense, so I skip the balance at import of users, and will handle it when importing transactions.

At some point we'll have to be able to clean up old transactions, and then such an 'initializing' (or rather consolidating) transaction would be handy too. But that's a different topic (worth the creation of another issue?).

Nevertheless, I'm still curious to understand how the balance field is ignored at insert. I understand from your feedback it is intentional, but how is it configured?

**Comment by dimitrydhondt** [comment]
I will write a system test to cover the behaviour of PUT on /partyrelations. It should ignore the balance field. I cannot find any code that makes the balanced be ignored on PUT at this time. But it should ! :-)

**Comment by eeemarv** [comment]
Notice that it may be possible there are transactions in eLAS which have no corresponding user. Because it was possible before 2 years ago to delete users which were holding transactions. A solution is to recreate this users (with dummy data then).

**Comment by eeemarv** [comment]
Also notice some groups are RECYCLING LETSCODES. The implication is that you can't rely on the letscode to import the transactions.
The transactions table also doesn't hold letscodes. It only holds user-ids. Nevertheless, the csv-export of transactions in eLAS exports only letscodes, no user-ids. That's completely unreliable, as admins can change the letscode of a user (which happens in our group; the user gets another code when deactivated) or recycled (which also happens in our group). 
My suggestion is to use directly the database as the import source.

**Comment by eeemarv** [comment]
Also incomplete: the csv export of messages in eLAS. The "Description" field is missing.

**Comment by renardeau** [comment]
@eeemarv 
Sounds like your group is a good test case to validate our solution seen the several particularities :-)

From what I understand, the only way to guarantee uniqueness is to rely on PKs of the different elas tables. I suggest to prefix each PK with some group identifications, so that we can guarantee we're not importing same message/user/transaction twice.
I'll try matching all columns in inner-gerbil with equivalent columns in elas (should probably have been the first step for this task)
Description missing in elas message export sounds like a bug -> should be reported to elas project

**Comment by eeemarv** [comment]
[inner-elas mapping.xlsx](https://github.com/dimitrydhondt/inner-gerbil/files/57819/inner-elas.mapping.xlsx)

uploading the spreadsheet here @renardeau shared via the mailinglist.

**Comment by eeemarv** [comment]
Maybe better first the import the letsgroup itself and then the users? The info is in the 'config' table in eLAS.
